## Multiple Linear Regression with sklearn

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

from sklearn.linear_model import LinearRegression

In [2]:
data = pd.read_csv('1.02. Multiple linear regression.csv')
data.head()

SAT   GPA  Rand 1,2,3
0  1714  2.40           1
1  1664  2.52           3
2  1760  2.54           3
3  1685  2.74           3
4  1693  2.83           2

In [3]:
data.describe()

SAT        GPA  Rand 1,2,3
count    84.000000  84.000000   84.000000
mean   1845.273810   3.330238    2.059524
std     104.530661   0.271617    0.855192
min    1634.000000   2.400000    1.000000
25%    1772.000000   3.190000    1.000000
50%    1846.000000   3.380000    2.000000
75%    1934.000000   3.502500    3.000000
max    2050.000000   3.810000    3.000000

'Rand 1,2,3' randomly assigned 1,2, or 3 to each sample

### Create Regression

In [4]:
x = data[['SAT','Rand 1,2,3']] # independent
y = data['GPA'] # dependent

In [5]:
reg = LinearRegression()
reg.fit(x,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

#### coefficient of 'SAT' and 'Rand 1,2,3', respectively

In [6]:
reg.coef_

array([ 0.00165354, -0.00826982])

#### intercept

In [7]:
reg.intercept_

0.29603261264909486

#### R-squared

In [8]:
reg.score(x,y)

0.4066811952814285

### Formula for Adjusted R^2

$R^2_{adj.} = 1 - (1-R^2)*\frac{n-1}{n-p-1}$

In [9]:
x.shape

(84, 2)

n = 84(=number of observations). p = 2(=number of predictors)

In [11]:
r2 = reg.score(x,y)

n = x.shape[0] # 84

p = x.shape[1] # 2

adjusted_r2 = 1-(1-r2)*(n-1)/(n-p-1)
adjusted_r2

0.39203134825134023

## Feature Selection

simplified models, improves speed and prevents a series of unwanted issues arising from having too many features

NOTE) if a variable has a p-value > 0.05, we can disregard it

### F_regression

creates simple linear regressions of each feature('SAT' and 'Rand 1,2,3') and the dependent variable

NOTE) for a simple linear regression, the p-value of F-stat is equivalent to the p-value of the only independent variable

In [12]:
from sklearn.feature_selection import f_regression

In [13]:
f_regression(x,y)

(array([56.04804786,  0.17558437]), array([7.19951844e-11, 6.76291372e-01]))

The first array contains the F-statistics for each of the regressions

The second array is corresponding p-values

In [14]:
p_values = f_regression(x,y)[1]
p_values

array([7.19951844e-11, 6.76291372e-01])

In [15]:
# round to 3 digits after the dot
p_values.round(3)

array([0.   , 0.676])

#### p-value for 'SAT' = 0.000

#### p-value for 'Rand 1,2,3' = 0.676

We can determine that 'Rand 1,2,3' is not a useful variable.

## Create a summary table

In [19]:
reg_summary = pd.DataFrame(data=x.columns.values, columns = ['Features'])
# x.columns.values = 'SAT' and 'Rand 1,2,3'
reg_summary

Features
0         SAT
1  Rand 1,2,3

In [20]:
reg_summary['Coefficients'] = reg.coef_
reg_summary['p-values'] = p_values.round(3)

In [21]:
reg_summary

Features  Coefficients  p-values
0         SAT      0.001654     0.000
1  Rand 1,2,3     -0.008270     0.676